utils/rand/rand.go

```diff
@@ -0,0 +1,169 @@
+	// get the bit size of max
+	mask := uint64(0)
+	for max&mask != max {
+		mask = (mask << 1) | 1
+	}
+
+	// Using 64 bits of random and reducing modulo n does not guarantee a high uniformity
+	// of the result.
+	// For a better uniformity, loop till a sample is less or equal to `max`.
+	// This means the function might take longer time to output a random.
+	// Using the size of `max` in bits helps the loop end earlier (the algo stops after one loop
+	// with more than 50%)
+	// a different approach would be to pull at least 128 bits from the random source
+	// and use big number modular reduction by `n`.
+	random := n
+	for random > max {
+		if _, err := rand.Read(buffer[:size]); err != nil {
+			return 0, randFailure
+		}
+		random = binary.LittleEndian.Uint64(buffer)
+		random &= mask // adjust to the size of max in bits
+	}
+
+	return random, nil
+}
+
+// returns a random uint32
+func Uint32() (uint32, error) {
+	// for 64-bits machines, doing 64 bits operations and then casting
+	// should be faster than dealing with 32 bits operations
+	r, err := Uint64()
+	return uint32(r), err
+}
+
+// returns a random uint32 strictly less than n
+// errors if n==0
+func Uint32n(n uint32) (uint32, error) {
+	if n == 0 {
+		return 0, fmt.Errorf("n should be strictly positive, got %d", n)
+	}
+	// the max returned random is n-1 > 0
+	max := n - 1
+	// count the bytes size of max
+	size := 0
+	for tmp := max; tmp != 0; tmp >>= 8 {
+		size++
+	}
+	buffer := make([]byte, 4) // TODO: declare as a global variable and add a lock?
+	// get the bit size of max
+	mask := uint32(0)
+	for max&mask != max {
+		mask = (mask << 1) | 1
+	}
+
+	// Using 32 bits of random and reducing modulo n does not guarantee a high uniformity
+	// of the result.
+	// For a better uniformity, loop till a sample is less or equal to `max`.
+	// This means the function might take longer time to output a random.
+	// Using the size of `max` in bits helps the loop end earlier (the algo stops after one loop
+	// with more than 50%)
+	// a different approach would be to pull at least 128 bits from the random source
+	// and use big number modular reduction by `n`.
+	random := n
+	for random > max {
+		if _, err := rand.Read(buffer[:size]); err != nil {
+			return 0, randFailure
+		}
+		random = binary.LittleEndian.Uint32(buffer)
+		random &= mask // adjust to the size of max in bits
+	}
+
+	return random, nil
+}
+
+// returns a random uint
+func Uint() (uint, error) {
+	r, err := Uint64()
+	return uint(r), err
+}
+
+// returns a random uint strictly less than n
+// errors if n==0
+func Uintn(n uint) (uint, error) {
+	r, err := Uint64n(uint64(n))
+	return uint(r), err
+}
+
+// Shuffle permutes a data structure in place
+// based on the provided `swap` function.
+// It is not deterministic.
+//
+// It implements Fisher-Yates Shuffle using crypto/rand as a source of randoms.
+//
+// O(1) space and O(n) time
+func Shuffle(n uint, swap func(i, j uint)) error {
```

Comment:
The interface here mirrors the `math/rand` function https://pkg.go.dev/math/rand#Shuffle. 
Since this function is going to replace the `math/rand.Shuffle` in the repo, I simply chose the same parameters.